include/alpaka/mem/buf/BufCpu.hpp

```diff
@@ -307,6 +308,13 @@ namespace alpaka
         {
             using type = TIdx;
         };
+
+        template<typename TElem, typename TDim, typename TIdx>
+        struct MemVisibility<BufCpu<TElem, TDim, TIdx>>
+        {
+            using type = std::tuple<alpaka::MemVisibleCPU>;
+        };
+
```

Comment:
Actually, the memory in a `BufCpu` returned by a call to `allocMappedBuf` is visible by the devices in the corresponding platform.

But the C++ type is the same, so I don't know how it could be distinguished at compile time.

Reply:
I'm not sure, what you mean. 

Each time, if we use a device to allocate memory, the buffer will be accessible for the device. But afterwards you can try to access the memory with another device, e.g. a cuda device. If the memory was allocate with malloc and without HMM, the cuda device can never access the memory. If we would have HMM support, the type needs to be `std::tuple<alpaka::MemVisibleCPU, MemVisibleGpuCudaRt>`.

Reply:
Even without HMM support, if you allocate a host buffer with `allocMappedBuf`, it will be visible by the GPU.

However, the C++ type of the host buffer returned by `allocBuf` and `allocMappedBuf` is the same.

So, `trait::MemVisibility<TBuf>` cannot distinguish between the two cases.

Reply:
Ah, now I know what you mean. I missed `allocMappedBuf` two times :grimacing: First that it exist and second that you wrote `allocMappedBuf` and not `allocBuf`.

Yes your are right. Therefore I need to implement the visibility as template type of `BufCpu`, like I already did it for `ViewPlainPtr`:

https://github.com/alpaka-group/alpaka/blob/4054e15d844588b7b106b44d04c315de3191a9bd/include/alpaka/mem/view/ViewPlainPtr.hpp#L30-L38

I think the buffer type cannot have a default visibility type.